### ISSUE TITLE
Views have the wrong size on first load / reload

### ISSUE THREAD

**Issue by mjbvz** [microsoft/vscode#145082]
This tracks the issue discovered in https://github.com/microsoft/vscode-pull-request-github/issues/3227

See https://github.com/microsoft/vscode/issues/106585 for this discussion as well. We originally thought that issue had the same root causes as this one, but investigation revealed that was not the case

**Comment by mjbvz** [comment]
I've debugged this in an OSS build. It's a difficult consistently hitting the original issue from https://github.com/microsoft/vscode-pull-request-github/issues/3227 but I do some really odd behavior around view state being saved and restored

Here's what I was doing:

1. Open active PR view
2. Resize the views
3. Reload window  
4. Switch back to active PR view once it is loaded

After step 2, I see the following data being written for the views in [`updateWorkspaceState`](https://github.com/microsoft/vscode/blob/cf4fe193cf67bb6805433a839866e81825ecaf31/src/vs/workbench/services/views/common/viewContainerModel.ts#L111):

![Screen Shot 2022-03-14 at 4 15 30 PM](https://user-images.githubusercontent.com/12821956/158276560-1f2c0f17-04eb-451a-b1b3-e00120ee62ee.png)


However when the state is then restored by `initialize`, I see the following  data being loaded (look at the sizes specifically):

![Screen Shot 2022-03-14 at 4 18 05 PM](https://user-images.githubusercontent.com/12821956/158276594-47ae9b02-9d63-41b5-90c2-6344ad8f6a00.png)

Will try to track down if some other code is writing this incorrect size data or if the reload is somehow restoring it incorrectly  on reload

**Comment by mjbvz** [comment]
Adding @sbatten and @sandy081 to see if you have any ideas about what's going on here. I've been debugging through the storage calls (at least on the renderer side) and we seem be writing the values as expected but on reload we still get back the incorrect values

**Comment by mjbvz** [comment]
After debugging further, here's what is happening:

Basics:
- The GitHub PR view contains 3 views (from top to bottom): `prStatus:github`, `gitHub:activePullRequest`, `gitHub:activePullRequest:welcome`
- On reload with a PR, all 3 of these views start as visible.
- The PR extension then hides the `gitHub:activePullRequest:welcome` view by updating a context key

Bug:

1. Reload window
1. The  `github:activePullRequest:welcome` view is created first
1. After `SplitView.relayout` finishes, we have the sizes: `github:activePullRequest:welcome == 1012`
1. Next we add `prStatus:github`
1. After `SplitView.relayout` finishes, we have the sizes: `prStatus:github == 870, github:activePullRequest:welcome == 142`
1. Finally we add `gitHub:activePullRequest`
2. The sizes are now `prStatus:github == 750, github:activePullRequest == 142, github:activePullRequest:welcome == 120`
6. We now call `restoreViewSizes` for the various views (in a loop)
8. Here however, have very large view size for `gitHub:activePullRequest:welcome` (1012) 🐛 
9. This means that wen we do `restoreViewSizes` for the welcome view, the other two views to become quite small. In my testing after  `SplitView.layout` finishes here, the sizes of the three views are: `[142, 142, 728]`
10. We then remove the `gitHub:activePullRequest:welcome` view (I believe this is triggered by the PR extension changing the view's context key)
11. This causes the last view (`gitHub:activePullRequest`) to fill in the now empty space
12. After `SplitView.layout` complete this time, the sizes of the two remaining views are: `[142, 870]` 🐛 

The bug is probably that `gitHub:activePullRequest:welcome` saves off a very large view size from first load, and that we then try to incorrectly restore this large view size before the view is hidden

I don't know what the proper fix for this. Should we try saving off the fact that the welcome view has been conditionally hidden for example? Should we try saving a more accurate size for the welcome view? 

I think I've made about as much progress as I reasonable can in this space without any real understand of the view code. @sbatten @sandy081 @alexr00 Can you please take this issue from here?

**Comment by alexr00** [comment]
Thank you for looking into this @mjbvz and tracking it down 🙏.

**Comment by sandy081** [comment]
Sure @mjbvz - thanks for update

**Comment by sandy081** [comment]
To be clear, is this the bug we are planning to fix? (After refresh the bottom view is taking more space)

https://user-images.githubusercontent.com/10746682/159418304-d01f5d49-be1a-4ba1-9d12-39a5f7a38815.mov

**Comment by alexr00** [comment]
Yes, that's the one.

**Comment by sandy081** [comment]
I can reproduce this behaviour with following sample extension

<details><summary>Extension</summary>

```json
"viewsContainers": {
			"activitybar": [
				{
					"id": "github-pull-request",
					"title": "GitHub",
					"icon": "$(github)"
				}
			]
		}, 
		"views": {
			"github-pull-request": [
				{
					"id": "prStatus:github",
					"name": "Changes In Pull Request",
					"when": "github:inReviewMode",
					"icon": "$(git-pull-request)",
					"visibility": "visible"
				},
				{
					"id": "github:activePullRequest",
					"name": "Active Pull Request",
					"when": "github:inReviewMode && github:focusedReview"
				},
				{
					"id": "github:activePullRequest:welcome",
					"name": "Active Pull Request",
					"when": "!github:stateValidated && github:focusedReview"
				}
			]
		}
```

```ts
import * as vscode from 'vscode';

export async function activate(context: vscode.ExtensionContext) {
    vscode.window.createTreeView('prStatus:github', {
        treeDataProvider: {
            getChildren() {
                return [];
            },
            getTreeItem() {
                return new vscode.TreeItem('hi');
            }
        }
    });
    vscode.window.createTreeView('github:activePullRequest', {
        treeDataProvider: {
            getChildren() {
                return [];
            },
            getTreeItem() {
                return new vscode.TreeItem('hi');
            }
        }
    });
    vscode.window.createTreeView('github:activePullRequest:welcome', {
        treeDataProvider: {
            getChildren() {
                return [];
            },
            getTreeItem() {
                return new vscode.TreeItem('hi');
            }
        }
    });
    await vscode.commands.executeCommand('setContext', 'github:inReviewMode', true);
    await vscode.commands.executeCommand('setContext', 'github:focusedReview', true);
    await vscode.commands.executeCommand('setContext', 'github:stateValidated', false);
    setTimeout(() => {
        vscode.commands.executeCommand('setContext', 'github:stateValidated', true);
    }, 5000);
}
```

</details>

https://user-images.githubusercontent.com/10746682/159440804-98f3618c-8790-400a-ad58-188efdd3a872.mov

@mjbvz analysis is 💯 correct

In the beginning all 3 views are visible and are sized equal. After 5 seconds, last view is made hidden by setting the context. At this point, last view is simply removed and its space is allocated to the view (second view) before last view. First view takes 33% and Second view takes 67% of the total size. Instead the removed size shall be equally distributed to all visible views.

Used [DistributeSizing](https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/base/browser/ui/splitview/splitview.ts#L355 
) option while removing the pane and this seems to fix it.

https://user-images.githubusercontent.com/10746682/159441248-48ee06fb-6d6a-4e14-98a3-c349e32a8a28.mov

CC @sbatten @joaomoreno  - I do not have much experience in using split views and do not know what are the side affects of using `DistributeSizing` option. So please take a look and update/enhance the fix if needed.

**Comment by joaomoreno** [comment]
@mjbvz Great analysis!

---

@sandy081, the docs say:

> Whether to distribute other view's sizes.

![image](https://user-images.githubusercontent.com/22350/159461821-1347e328-6a1d-4ed7-bc60-4c534b9fa4ea.png)

Which means:

- When using `DistributeSizing` in `removeView`, all other views will be resized to distribute the total splitview size.
- Without that option, the regular layout algorithm will happen... usually that means that only the adjacent view will stretch.

**Comment by sandy081** [comment]
> When using DistributeSizing in removeView, all other views will be resized to distribute the total splitview size.

I am not sure if the following doc says the same or I am wrong:

https://github.com/microsoft/vscode/blob/4a130c40ed876644ed8af2943809d08221375408/src/vs/base/browser/ui/splitview/splitview.ts#L351-L355

As per the doc, I was expecting that the removed view space is distributed equally across other views. But it is not and it is doing as you mentioned. This also explains why the views sizes are getting reset to equal sizes after a view is removed (when DistributeSizing option was used).

**Comment by joaomoreno** [comment]
Thanks for calling that out! I'm improved the docs there. 👌

**Comment by sandy081** [comment]
Given that it is not doing what I am expecting - **removed view space is distributed equally across other views**. Is there an other option to achieve this?

**Comment by joaomoreno** [comment]
Not today, no.

**Comment by sandy081** [comment]
I think this behaviour is ok and acceptable when compared to existing and I also observe the same in editor groups.

https://user-images.githubusercontent.com/10746682/159555456-adfac500-dde2-4ade-922f-843673038843.mov

**Comment by sbatten** [comment]
I've approved the PR. I think the change we have made is a band-aid that helps, but doesn't cut to the root of the problem. I thought through various view/hide scenarios and in each case, I had a different feeling about what I expected to happen with the sizing of the views. IMO, our previous behavior was more correct from the pane view side as an intentional "hide" operation on a view should not redistribute every view in the container. However, the problem here is that there is a problematic extra view that is not controlled by the user. In this case, we are just trying to get sane sizing. If we invest more time in this, I think the fix lives at a higher layer.
